File: Kotlin/063.kt
import java.util.Scanner
import kotlin.math.max

fun main() {
	val input = Scanner(System.`in`)
	val H = input.nextInt()
	val W = input.nextInt()

	val P = Array(H) { IntArray(W) { input.nextInt() } }

	val stack = Array(H) { IntArray(W) }

	val count = IntArray(H * W)

	fun search(currentRowIndex: Int, assignedRows: Int): Int {
		if (currentRowIndex == H) {
			if (assignedRows == 0) return 0

			count.fill(0)
			for (x in stack[assignedRows - 1]) {
				if (x > 0) count[x - 1]++
			}

			return count.max()!! * assignedRows
		}

		val a = search(currentRowIndex + 1, assignedRows)

		if (assignedRows == 0) {
			P[currentRowIndex].copyInto(stack[0])
		}
		else {
			val prevStackRow = stack[assignedRows - 1]
			val currentStackRow = stack[assignedRows]

			for ((i, x) in P[currentRowIndex].withIndex()) {
				currentStackRow[i] = if (prevStackRow[i] == x) x else 0
			}
		}

		val b = search(currentRowIndex + 1, assignedRows + 1)

		return max(a, b)
	}

	println(search(0, 0))
}

File: Python/063.py
from sys import stdin
import numpy

H, W = map(int, stdin.readline().split())

P = numpy.fromstring(stdin.read(), dtype=int, sep=' ').reshape(H, W)

stack = numpy.ndarray((H, W), dtype=int)

def search(currentRowIndex: int, assignedRows: int) -> int:
	if currentRowIndex == H:
		if assignedRows == 0:
			return 0

		count = numpy.bincount(stack[assignedRows - 1])

		if len(count) == 1:
			return 0

		return count[1:].max() * assignedRows

	a = search(currentRowIndex + 1, assignedRows)

	if assignedRows == 0:
		stack[0] = P[currentRowIndex]
	else:
		prevStackRow = stack[assignedRows - 1]
		stack[assignedRows] = numpy.where(
			prevStackRow == P[currentRowIndex],
			prevStackRow,
			0)

	b = search(currentRowIndex + 1, assignedRows + 1)

	return max(a, b)

print(search(0, 0))
